module/zfs/mmp.c

```diff
@@ -125,27 +130,79 @@ mmp_thread_stop(dsl_pool_t *dp)
 	mmp->mmp_thread_exiting = 0;
 }
 
+vdev_t *
+vdev_random_leaf(spa_t *spa)
+{
+	vdev_t *vd, *child;
+
+	ASSERT(spa);
+	ASSERT(spa_config_held(spa, SCL_STATE, RW_READER) == SCL_STATE);
+
+	vd = spa->spa_root_vdev;
+	while (!vd->vdev_ops->vdev_op_leaf) {
+		child = vd->vdev_child[spa_get_random(vd->vdev_children)];
+		if (!vdev_is_dead(child))
+			vd = child;
+	}
+
+	return (vd);
+}
+
 static void
 mmp_thread(dsl_pool_t *dp)
 {
 	mmp_thread_state_t *mmp = &dp->dp_mmp;
-	clock_t start;
-	/*
 	spa_t *spa = dp->dp_spa;
-	*/
+	abd_t *abd;
+	char *buf;
+	int bufsize = (1ULL << MAX_UBERBLOCK_SHIFT);
+
+	abd = abd_alloc(bufsize, B_TRUE);
+	buf = kmem_alloc(bufsize, KM_SLEEP);
+	bzero(buf, bufsize);
 
 	mmp_thread_enter(mmp);
 
 	dprintf("entered: pool %p\n", dp);
 
 	mmp->mmp_last_update = gethrtime();
 	for (;;) {
+		int flags = ZIO_FLAG_CONFIG_WRITER | ZIO_FLAG_CANFAIL;
+		clock_t start;
+		int l, n, rc;
+		vdev_t *vd;
+		zio_t *zio;
+
 		if (mmp->mmp_thread_exiting)
-			mmp_thread_exit(mmp, &mmp->mmp_thread);
+			goto cleanup_and_exit;
 
 		start = gethrtime();
 
-		/* Do actual mmp reads/writes here */
+		/* Copy the latest uberblock and update the time */
+		(void) memcpy(buf, &spa->spa_ubsync, sizeof (uberblock_t));
+		((uberblock_t *)buf)->ub_timestamp = gethrestime_sec();
```

Comment:
Generally, I agree with you.  However, the case of multiple nodes sharing storage, which are running different versions of ZFS, is a case I've specifically dropped from the simplified MMP design.  

My rationales are:
1. I believe it's a very small fraction of systems which share storage.  Most of what I hear about falls into the "bought an appliance" category or the "bought a system from a vendor and treat it like an appliance".  In either case the vendor would not likely set up dissimilar versions.
2. If this situation is set up by a user unaware of MMP, the default values will be in place.  If we change the default values in a future ZFS release, we will know the prior defaults and can set the new values to interact correctly.
3. There are other similar potential problems Simplified MMP does not attempt to solve.  For example, an older non-MMP-aware ZFS can force import a pool which is already imported by a node running an MMP-aware ZFS.  Simplified MMP does not detect that.  Similarly, there is no protection against adding device(s) to a pool when they are already in use, in a pool imported on another node.  Such situations could be detected, but that is left for enhancement in order to get the main case covered soon enough to be helpful.

Do you disagree strongly with those?